src/SonarScanner.MSBuild.PreProcessor/JreCaching/TarGzUnpacker.cs

```diff
@@ -0,0 +1,94 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using ICSharpCode.SharpZipLib.Core;
+using ICSharpCode.SharpZipLib.GZip;
+using ICSharpCode.SharpZipLib.Tar;
+using SonarScanner.MSBuild.Common;
+
+namespace SonarScanner.MSBuild.PreProcessor.JreCaching;
+
+public class TarGzUnpacker(IDirectoryWrapper directoryWrapper, IFileWrapper fileWrapper, IOperatingSystemProvider operatingSystemProvider) : IUnpacker
+{
+    // ref https://github.com/icsharpcode/SharpZipLib/blob/ff2d7c30bdb2474d507f001bc555405e9f02a0bb/src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs#L608
+    public void Unpack(Stream archive, string destinationDirectory)
+    {
+        using var gzipStream = new GZipInputStream(archive);
+        using var tarIn = new TarInputStream(gzipStream, null);
+
+        var destinationFullPath = Path.GetFullPath(destinationDirectory).TrimEnd('/', '\\');
+        while (tarIn.GetNextEntry() is {} entry)
+        {
+            if (entry.TarHeader.TypeFlag is not (TarHeader.LF_LINK or TarHeader.LF_SYMLINK))
+            {
+                ExtractEntry(tarIn, destinationFullPath, entry);
+            }
+        }
+    }
+
+    // ref https://github.com/icsharpcode/SharpZipLib/blob/ff2d7c30bdb2474d507f001bc555405e9f02a0bb/src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs#L644
+    private void ExtractEntry(TarInputStream tar, string destinationFullPath, TarEntry entry)
+    {
+        var name = entry.Name;
+
+        if (Path.IsPathRooted(name))
+        {
+            // NOTE:
+            // for UNC names...  \\machine\share\zoom\beet.txt gives \zoom\beet.txt
+            name = name.Substring(Path.GetPathRoot(name).Length);
+        }
+
+        name = name.Replace('/', Path.DirectorySeparatorChar);
+
+        var destinationFile = Path.Combine(destinationFullPath, name);
+        var destinationFileDirectory = Path.GetDirectoryName(Path.GetFullPath(destinationFile)) ?? string.Empty;
+
+        var isRootDir = entry.IsDirectory && entry.Name == string.Empty;
+
+        if (!isRootDir && !destinationFileDirectory.StartsWith(destinationFullPath, StringComparison.InvariantCultureIgnoreCase))
+        {
+            throw new InvalidNameException("Parent traversal in paths is not allowed");
+        }
+
+        if (entry.IsDirectory)
+        {
+            directoryWrapper.CreateDirectory(destinationFile);
+        }
+        else
+        {
+            directoryWrapper.CreateDirectory(destinationFileDirectory);
+
+            using var outputStream = fileWrapper.Create(destinationFile);
+            tar.CopyEntryContents(outputStream);
+
+#if NETSTANDARD
+            if (operatingSystemProvider.IsUnix())
```

Comment:
We may have a problem on Apline and this is how others are solving it:

https://github.com/xamarin/XamarinComponents/issues/324

https://sourcegraph.com/github.com/Jackett/Jackett/-/blob/src/Jackett.Server/Services/FilePermissionService.cs?L11-57